pkg/events/sorting/queue.go

```diff
@@ -0,0 +1,107 @@
+package sorting
+
+import (
+	"fmt"
+	"github.com/aquasecurity/tracee/pkg/external"
+	"sync"
+)
+
+type eventNode struct {
+	event       *external.Event
+	previous    *eventNode
+	next        *eventNode
+	isAllocated bool
+}
+
+// A double linked list used to store events in LIFO order
+type eventsQueue struct {
+	pool  eventsPool
+	tail  *eventNode
+	head  *eventNode
+	mutex sync.Mutex
+}
+
+// Put insert new event to the double linked list
+func (eventsQueue *eventsQueue) Put(newEvent *external.Event) {
+	newNode := eventsQueue.pool.Alloc(newEvent)
+	eventsQueue.mutex.Lock()
+	defer eventsQueue.mutex.Unlock()
+	eventsQueue.put(newNode)
+}
+
+// Get remove the node at the head of the queue and return it
+// Might return error with a valid event in case of internal pool error, for the user to know that an error occured
+// but was contained
+func (eventsQueue *eventsQueue) Get() (*external.Event, error) {
+	eventsQueue.mutex.Lock()
+	defer eventsQueue.mutex.Unlock()
+	if eventsQueue.head == nil {
+		if eventsQueue.tail != nil {
+			return nil, fmt.Errorf("unmatching values of head and tail - one should be nil only if queue is empty")
+		}
+		return nil, nil
+	}
+	headNode := eventsQueue.head
+	if headNode == eventsQueue.tail {
+		if headNode.next != nil || headNode.previous != nil {
+			return nil, fmt.Errorf("trying to remove the only node in queue, but it is connected to another node")
+		}
+		eventsQueue.tail = nil
+		eventsQueue.head = nil
+	} else {
+		if headNode.previous == nil {
+			return nil, fmt.Errorf("a node which is not the tail without previous is given to remove")
+		}
+		if headNode.next != nil {
+			return nil, fmt.Errorf("head event to remove has next")
+		}
+		headNode.previous.next = nil
+		eventsQueue.head = headNode.previous
+	}
+	headNode.previous = nil
+	headNode.next = nil
+	extractedEvent := headNode.event
+	err := eventsQueue.pool.Free(headNode)
```

Comment:
The only possible error here is a call to Free() when there are no allocated eventNode's. If that is the case, why not to simply deal with that silently, instead of propagating this error to upper calls ?